### ISSUE TITLE
How to estimate bpp of a standalone tensor using EntropyBottleneck() layer?

### ISSUE THREAD

**Issue by Somdyuti2** [InterDigitalInc/CompressAI#62]
I want to estimate the bpp of a tensor without compressing it, and tried to use the `EntropyBottleneck` layer of CompressAI for this purpose. However, when I used it on a 512 x 512  image tensor and a random tensor, I got nearly the same bpp value of around 20 with an untrained `EntropyBottleneck` layer in either case. Theoretically, the **random tensor should have much higher bpp than the image**, and also, since the pixel values for both the tensors lie between 0-255, the **bpp should be no more than 8 even for a fully randomized tensor** (I am not sure if the data type of `torch.float32` plays a role here).  However, this was an **untrained layer**, and naturally, I thought that the bpps returned are just random values  and training the layer would fix the issue, and given me an close estimate of the true bpp. But even after thousands of iterations, the bpp values for both tensors **did not change from their initial value** (only the 3rd and 4th digits after decimal fluctuated, without a  monotonic increase or decrease), **even though the entropy bottleneck loss steadily decreased**. 

Here is a minimal working example of the issue. I tried to closely follow this [tutorial](https://interdigitalinc.github.io/CompressAI/tutorials/tutorial_custom.html) for custom model, without the encoder and decoder layers (as I do not need to compress the tensor).  I am just showing the example with the image tensor to keep it short. The random tensor has the same behavior. 

```
from skimage import data
import math

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

from compressai.entropy_models import EntropyBottleneck
from compressai.models import CompressionModel

# Define a network having only the EntropyBottleneck layer
class Network(CompressionModel):
    def __init__(self, N=128):
        super().__init__(N) # using super().__init__() as given in the tutorial resulted in an error: __init__() missing 1 required positional argument: 'entropy_bottleneck_channels'
        self.entropy_bottleneck = EntropyBottleneck(N)
        

    def forward(self, x):
        
        x_hat, x_likelihoods = self.entropy_bottleneck(x)
        return x_hat, x_likelihoods

# initialize the network with entropy bottleneck layer with only 1 channel for a single grayscale image
net = Network(N=1) 

# read image data and convert to tensor
image = data.camera() # read cameraman image
H, W = image.shape
img_tensor = torch.tensor(image, dtype=torch.float32).view(-1, 1, H, W)
# rand_tensor = torch.randint(0,255,(1,1,H,W), dtype=torch.float32) 
N, _, H, W = img_tensor.size()
num_pixels = N * H * W

# compute bpp with untrained layer
img_approx, img_likelihood = net(img_tensor)
bpp_loss_img = torch.log(img_likelihood).sum() / (-math.log(2) * num_pixels)
print ("Image BPP", bpp_loss_img)

# Train EntropyBottleneck layer

aux_parameters = set(p for n, p in net.named_parameters() if n.endswith(".quantiles"))
aux_optimizer = optim.Adam(aux_parameters, lr=1e-3)
for i in range(1000):
    aux_optimizer.zero_grad()
    img_hat, img_likelihoods = net(img_tensor)
    aux_loss = net.aux_loss()
    aux_loss.backward()
    aux_optimizer.step()
    bpp_loss_img = torch.log(img_likelihoods).sum() / (-math.log(2) * num_pixels)
    print(bpp_loss_img, aux_loss)
```

I also tried using an `EntropyBottleneck` by itself, i.e. without the custom network inheriting `CompressionModel`, but the outcome was the same. It will be very helpful if you could help me debug this issue and/or let me know how can I estimate the bpp of a standalone tensor using this library. Thanks.

**Comment by jbegaint** [comment]
you still have to learn the main parameters of the EntropyBottleneck. That's why your bpp_loss is not decreasing. The slight difference in values is due to the additive uniform noise in the forward pass.

Now I'm not sure to understand what you are trying to do, but the EntropyBottleneck module has a fairly simple modeling of the samples distribution that is probably not suitable for raw pixel data. If you don't want to compress your image then you might want to look at lossless end-to-end compression network.

**Comment by Somdyuti2** [comment]
@jbegaint  thanks for the reply. I think I understand the error now. From the compressAI custom model tutorial in the documentation that I followed, it seemed like that the auxiliary parameters were referring to the `EntropyBottleneck`s parameters, and the main parameters were just the parameter's of the layers in the encoder and decoder. However, that was wrong and the bpp loss decreased after also including the  set of  `EntropyBottleneck`s  parameters which satisfy ` not n.endswith(".quantiles")` following your reply. 

Also, thank you for the additional information about lossless end-to-end compression network. Indeed it might be useful to me.
